src/ratbag_emu/protocol/base.py

```diff
@@ -11,24 +11,10 @@
 from ratbag_emu.protocol.util.profile import Profile
 
 
-class BaseDevice(UHIDDevice):
+class SimulatedDevice(UHIDDevice):
     verbose = True
 
-    '''
-    Represents an hardware device
-    '''
-
-    protocol = None
-
-    '''
-    Hardware profile
-
-    Represents the active profile in the hardware.
-    '''
-    hprof = Profile()
-
-    active_profile = None
-    profiles = []
+    hprof = None
```

Comment:
> why have you removed the documentation?

What do you mean?

> are you sure this is a good idea to have a class attribute which seems to be instance specific?

I don't think that will generate any issues because it is not shared between devices. I create a profile instance per emulated device. Actually, I want that to be shared between the UHID devices of an emulated device.

Reply:
I think there is a rule that "hardware" has to be only ever abbreviated with "hw" anyway. And if there isn't there should totally be one :)

Reply:
> are you sure this is a good idea to have a class attribute which seems to be instance specific?

this is a bit dangerous because python's instance vs class attributes are a bit weird and there's a risk of bus because it won't do what you'd expect:
```
>>> class Foo:
...     bar = None
...     def __init__(self, n):
...             self.bar = n
... 
>>> f = Foo('foo')
>>> b = Foo('bar')
>>> f.bar
'foo'
>>> b.bar
'bar'
>>> Foo.bar
# None
```

IOW, class attributes must be either accessed by class name only or not overwritten in the class with `self`, otherwise they become instance attributes - without changing the class attribute.

This is a python quirk and the general rule I found is "don't declare variables outside `__init__()`". It's very different to other languages (well, Java for example)

Given how you use the `hprof` below, you **definitely** should not declare it as class variable but set it to `None` inside `__init__()`

Reply:
> please use correct names, we are not limited and hprof is not straightforward as hardware_profile. Also, this has the good property of being able to append an s at the end, to tell us if this is a collection or a unique object.

Changed to `hw_profile` as Peter suggested. This variable represents the active settings on the device. The profile list is handled in a different variable (`profiles`). I did this because in some brands (specifically Logitech) the active profile might not be the hardware profile. When you change a profile, the profile settings will be copied to the hardware profile. You can then change a setting directly (ex. using AdjutableDPI). This won't change the active profile, just the hardware profile.

> why have you removed the documentation?

I moved it to `BaseDevice`.

> are you sure this is a good idea to have a class attribute which seems to be instance specific?

I changed that. After I finish dealing with everything I will push the changes for re-review.

Reply:
> Changed to hw_profile as Peter suggested. This variable represents the active settings on the device. The profile list is handled in a different variable (profiles). I did this because in some brands (specifically Logitech) the active profile might not be the hardware profile. When you change a profile, the profile settings will be copied to the hardware profile. You can then change a setting directly (ex. using AdjutableDPI). This won't change the active profile, just the hardware profile.

This distinction between hardware profile, active profile and list of profiles seems too tight to a specific vendor.
From what you said, it seems the `hw_profile` is actually the `active_profile`.
This base class is an ideal view of a gaming mouse, and any specifics should be handled in the various protocols.

IMO, it would make more sense to have a list of profiles, and an active profile index, in the same way libratbag exports its profiles.

>>  why have you removed the documentation?
>
> I moved it to BaseDevice.

If it's in a different class, you have to duplicate the documentation. Because you will not inherit it from the other class.

Reply:
> This distinction between hardware profile, active profile and list of profiles seems too tight to a specific vendor.
From what you said, it seems the hw_profile is actually the active_profile.
This base class is an ideal view of a gaming mouse, and any specifics should be handled in the various protocols.
IMO, it would make more sense to have a list of profiles, and an active profile index, in the same way libratbag exports its profiles.

We do have that, we just have this extra field.